onnxruntime/core/providers/dml/DmlExecutionProvider/src/Operators/DmlDFT.h

```diff
@@ -374,6 +228,69 @@ class GpuDFTOperator : public WRL::Base<IMLOperatorKernel>
             context->GetExecutionInterface(executionObject.GetAddressOf());
             executionObject.As(&commandList);
 
+            auto dftParams = PrepareDFT(context);
+            switch (dftParams.Type)
+            {
+                case DFTType::Stockham:
+                {
+                    StockhamFFT(dftParams, commandList.Get());
+                    break;
+                }
+                case DFTType::BluesteinZChirp:
+                    __fallthrough;
+                default:
+                    return E_NOTIMPL;
+            }
+        }
+        catch (...)
+        {
+            return E_FAIL;
+        }
+
+        return S_OK;
+    }
+
+    DFTParameters PrepareDFT(IMLOperatorKernelContext* context)
+    {
+        DFTParameters params = {};
+
+        // Get the input and output shape sizes
+        ComPtr<IMLOperatorTensor> inputTensor;
+        ORT_THROW_IF_FAILED(context->GetInputTensor(0, &inputTensor));
+        auto inputDims = GetTensorDimensions(inputTensor.Get());
+        auto inputDataSize = ComputeElementCountFromDimensions(inputDims);
+        ML_CHECK_VALID_ARGUMENT(static_cast<size_t>(m_axis) < inputDims.size())
+
+        ComPtr<IMLOperatorTensor> outputTensor;
+        ORT_THROW_IF_FAILED(context->GetOutputTensor(0, &outputTensor));
+        auto outputDims = GetTensorDimensions(outputTensor.Get());
+        auto outputDataSize = ComputeElementCountFromDimensions(outputDims);
+
+        ORT_THROW_HR_IF(E_FAIL, inputDims.size() != outputDims.size());
+
+        // Get optional dft_length input
+        ComPtr<IMLOperatorTensor> dftLengthTensor;
+        if (SUCCEEDED(context->GetInputTensor(1, &dftLengthTensor)) && dftLengthTensor != nullptr)
+        {
+            MLOperatorTensor tensor(dftLengthTensor.Get());
+            params.DFTLength = gsl::narrow_cast<uint32_t>(OperatorHelper::ReadScalarTensorCastToInt64(tensor));
+        }
+        else
+        {
+            params.DFTLength = inputDims[m_axis];
+        }
+
+        auto isPowerOfTwo = [](uint32_t n) { return (n != 0) && ((n & (n - 1)) == 0); };
+        if (!isPowerOfTwo(params.DFTLength))
+        {
+            params.Type = DFTType::BluesteinZChirp;
+            params.StockhamParams = {};
```

Comment:
why is this implemented but about BluesteinZChirp returns E_NOTIMPL?

Reply:
BluestainZChirp is used for non-powers of 2 DFT which is not implemented atm.